apt-hook/hook.cc

```diff
@@ -201,69 +282,230 @@ int main(int argc, char *argv[])
-                           "%d additional updates are available with UA Infra: ESM.",
-                           res.disabled_esms_i),
-                  res.disabled_esms_i);
-         ioprintf(std::cout, "\n");
+         if (res.enabled_esms_i > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d of the updates is from UA Infra: ESM.",
+                              "%d of the updates are from UA Infra: ESM.",
+                              res.enabled_esms_i),
+                     res.enabled_esms_i);
+            ioprintf(std::cout, "\n");
+         }
+         if (res.enabled_esms_a > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d of the updates is from UA Apps: ESM.",
+                              "%d of the updates are from UA Apps: ESM.",
+                              res.enabled_esms_a),
+                     res.enabled_esms_a);
+            ioprintf(std::cout, "\n");
+         }
       }
-      if (res.disabled_esms_a > 0)
+
+      if (res.disabled_esms_i > 0 || res.disabled_esms_a > 0)
       {
-         ioprintf(std::cout,
-                  ngettext("%d additional update is available with UA Apps: ESM.",
-                           "%d additional updates are available with UA Apps: ESM.",
-                           res.disabled_esms_a),
-                  res.disabled_esms_a);
+         if (command_used != "update")
+            std::cout << std::endl;
+         if (res.disabled_esms_i > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d additional update is available with UA Infra: ESM.",
+                              "%d additional updates are available with UA Infra: ESM.",
+                              res.disabled_esms_i),
+                     res.disabled_esms_i);
+            ioprintf(std::cout, "\n");
+         }
+         if (res.disabled_esms_a > 0)
+         {
+            ioprintf(std::cout,
+                     ngettext("%d additional update is available with UA Apps: ESM.",
+                              "%d additional updates are available with UA Apps: ESM.",
+                              res.disabled_esms_a),
+                     res.disabled_esms_a);
+            ioprintf(std::cout, "\n");
+         }
+
+         ioprintf(std::cout, gettext("To see these additional updates run: apt list --upgradable"));
+         ioprintf(std::cout, "\n");
+         ioprintf(std::cout, gettext("See https://ubuntu.com/advantage or run: sudo ua status"));
          ioprintf(std::cout, "\n");
       }
-
-      ioprintf(std::cout, gettext("To see these additional updates run: apt list --upgradable"));
-      ioprintf(std::cout, "\n");
-      ioprintf(std::cout, gettext("See https://ubuntu.com/advantage or run: sudo ua status"));
-      ioprintf(std::cout, "\n");
+   } else if (subcommand == PreInvoke) {
+      if (command_used == "upgrade" || command_used == "dist-upgrade") {
```

Comment:
Here is the main new message output functionality.

Note the different files it is currently looking at and in what situations. Let me know if this doesn't jive with your understanding of what this should be doing.